### ISSUE TITLE
rate button not disbaled for anonymous user 

### ISSUE THREAD

**Issue by MagdaN** [liqd/adhocracy3.mercator#1347]
when i am not logged in the rate button is not disabled in result phase

**Comment by xi** [comment]
The desired situation is something like this:

-   If the user is not logged in, we show functionality available to "standard" users. If users try to use the functionality, they are redirected to login.
-   If the user is logged in, they see the actual availability of the functionality.
-   If a feature is not available, it is either marked as disabled or is hidden completely.

The implementation of this is not very consistent and not very complete. Often, a feature is visible by default. In #1164 I implemented a method to disable features if a user is logged in but does not have the required permission. This does not cover the case where a user is not logged in and a "standard" user would not have the permission.

For this, we need an extension of that backend API that allows us to get permissions of a "standard" user.

I talked to @rukala about this and we decided to not fix this now and do a proper fix in one of the upcoming sprints.

**Comment by MagdaN** [comment]
Should we not at least hide it than also based on the phase?

**Comment by xi** [comment]
Today I thought a bit more about this and collected some issues connected to this:

1.   We use `cameFrom` to get back after redirecting to login. So #1137 and #1061 apply.
2.   If the functionality that is shown to the anonymous user has state (e.g. `textarea`) that state is lost after redirecting to login.
3.  The frontend does not know the permissions of a "standard" user.
   -  The current workaround for this is to hardcode / guess the permissions. This approach works reasonably well for special cases but is not applicable for reusable features (e.g. comments).
4.   After successful login, should the requested action be automatically executed or should I just be redirected to the same place?
5.   Maybe related: If functionality is visible in the frontend even though there is no permission, this will result in backend errors. Unfortunately, there is no concept for showing useful messages on backend errors. Even the error parsing is sometimes broken (e.g. #748)

(1) is a frontend bug tracked in separate issues. (2) can be mitigated by replacing the actual functionality with a button for anonymous users. (3) can be fixed by extending the API as described above. (4) is an open question and (5) is IMHO only related to specific workarounds.

So I guess if we use this issue to track (3) and the respective issues to track (1) this can all be fixed.

**Comment by pallix** [comment]
Can we remove backend tag on this?

**Comment by xi** [comment]
> Can we remove backend tag on this?

No. In order for this to be fixed the API needs to be extended (see (3) from my previous comment).

**Comment by joka** [comment]
Just to clear some responsibilities regarding the backend, it:

- cannot say what the user can do until the user is logged in (we don't know about local roles, groups)

- should not tell the frontend what buttons to show (this is part of frontend customization/config,
      also there is no generic solution)

- can say what a specific role can do

So I propose the following API Extensions:

- Extend Options response with mapping  `roles` repeating  the normal data structure for every user role:
  
   {'roles': {participator`: {'PUT', {request_body: ...

- The extendend option response is only send for anonymous request and only for 
resources with assigned workflow.
(Options request are already noisy and quite costly, we should not make it worse)

- The `standard` user is equal to the 'participator' role.
  (This is True for all workflows as far as we can see)

At this opportunity to reduce noise we should remove data from option responses that is actually not  not needed, like `request_body`, `response_body`.

**Comment by xi** [comment]
I agree completely. The only thing where I am not convinced is the exact API: Wouldn't it be better to keep the OPTIONS response as it is, but allow to specify a role via query parameter?

**Comment by joka** [comment]
Ok, its just  bit more work (we need make sure caching works)
What about the parameter name, `add_roles=True` ?

**Comment by joka** [comment]
.. not to forget the option request to get the query parameter for the options request

**Comment by joka** [comment]
The downside is adding query parameter is quite unusal, also we have to remove it if we want to get rid of options request to improve performance.

**Comment by joka** [comment]
mmh and it is not nice to mix get query parameter with options parameter

**Comment by xi** [comment]
> What about the parameter name, `add_roles=True` ?

I was thinking of `role=participant`. The important thing for me is that the structure of the response is always the same.

**Comment by joka** [comment]
So you want to make two different requests, one for the  participator role, one for the current user?

**Comment by joka** [comment]
`options  /process`  returns the options for the current user
`options /process/?role=<role>` returns the options for the specific role

**Comment by xi** [comment]
Yes. This has the benefit that the structure of the response is fully backwards compatible.
